### PR TITLE
Emulate

### PR DESCRIPTION
**Pull request by immunity-anomaly**
Emulate now working for all commands.

### CODE REVIEW

File: vosekast_control/TestSequence.py [hunk 1]
```diff
@@ -31,134 +32,205 @@ def __init__(
...
+                        self.scale.scale_history[0] - scale_nulled, 5)
+
+                    # #write values to csv file
+                    # with open('sequence_values.csv', 'a', newline='') as file:
+                    #     writer = csv.writer(file, delimiter=',',
+                    #                     quotechar='|', quoting=csv.QUOTE_MINIMAL)
+                    #     writer.writerow([self.scale.scale_history[1], scale_actual, self.scale.flow_history[0], flow_average])
+
+                    # db
+                    dbconnect = sqlite3.connect('sequence_values.db')
```
Comment (karlTGA) [inline vosekast_control/TestSequence.py]: please add a todo comment here, that we have to move the db init in a class or a singleton

File: vosekast_control/TestSequence.py [hunk 2]
```diff
@@ -31,134 +32,205 @@ def __init__(
...
+                    self.vosekast.measuring_tank_switch.close()
+                    self.vosekast.measuring_tank.drain_tank()
+                    self.logger.debug(
+                        "Draining measuring tank, opening Measuring Tank bypass.")
+
+            except:
+                self.logger.warning("Write loop killed, stopping sequence.")
+                dbconnect.close()
+                await self.stop_sequence()
+        else:
```
Comment (karlTGA) [inline vosekast_control/TestSequence.py]: same here. no unnecessary code copy and paste.

File: vosekast_control/Scale.py [hunk 3]
```diff
@@ -124,25 +119,44 @@ def start_measurement_thread(self):
...
+            self.logger.info("constant_tank_ready: " +
+                            str(self.vosekast.constant_tank.is_filled))
+            self.logger.info("measuring_tank_ready: " + str(self.vosekast.measuring_drain_valve.is_closed
+                                                            and not self.vosekast.measuring_tank.is_filled))
+            self.logger.info("constant_pump_running: " +
+                            str(self.vosekast.pump_constant_tank.is_running))
+            self.logger.info("measuring_drain_valve.is_closed: " +
+                            str(self.vosekast.measuring_drain_valve.is_closed))
+            self.logger.info("measuring_tank.is_filled: " +
+                            str(self.vosekast.measuring_tank.is_filled))
```
[the remark this answers is not in the document]
Comment (immunity-anomaly) [reply vosekast_control/Scale.py]: ok.

File: vosekast_control/Tank.py [hunk 4]
```diff
@@ -119,6 +121,29 @@ def prepare_to_fill(self):
...
+
+                    self.logger.debug(
+                        str(delta_time_filling) + 's < time allotted (90s)')
+                    await asyncio.sleep(1)
+                
+                return
+            except:
+                self._state = self.STOPPED
+                self.logger.warning("Filling {} aborted.".format(self.name))
+                return
```
[the remark this answers is not in the document]
Comment (immunity-anomaly) [reply vosekast_control/Tank.py]: ok. will be fixed in next commit

File: vosekast_control/TestSequence.py [hunk 5]
```diff
@@ -31,134 +32,205 @@ def __init__(
...
-            # todo if already full
 
-            # await constant_tank full
+        if not self.state == self.WAITING and not self.state == self.MEASURING:
+            try:
+                self.logger.info("Initialising sequence.")
+
+                # change state
+                self.state = self.WAITING
+                self.vosekast.state = self.vosekast.PREPARING_MEASUREMENT
```
[the remark this answers is not in the document]
Comment (immunity-anomaly) [reply vosekast_control/TestSequence.py]: tank/fill function already calls a method that changes vosekast state. removed duplicate.

File: vosekast_control/TestSequence.py [hunk 6]
```diff
@@ -31,134 +32,205 @@ def __init__(
...
+
+                # check if already running
+                if self.scale.is_running != True:
+                    self.scale.open_connection()
+                    self.scale.start_measurement_thread()
```
[the remark this answers is not in the document]
Comment (immunity-anomaly) [reply vosekast_control/TestSequence.py]: added method in scale that calls open_connection and start_measurement_thread.

File: vosekast_control/TestSequence.py (hunk 1 above)
Comment (immunity-anomaly) [reply vosekast_control/TestSequence.py]: ok

File: vosekast_control/TestSequence.py (hunk 2 above)
Comment (immunity-anomaly) [reply vosekast_control/TestSequence.py]: removed unnecessary code.

File: vosekast_control/Scale.py [hunk 7]
```diff
@@ -119,45 +123,28 @@ def start_measurement_thread(self):
...
-                                                            and not self.vosekast.measuring_tank.is_filled))
-            self.logger.info("constant_pump_running: " +
-                            str(self.vosekast.pump_constant_tank.is_running))
-            self.logger.info("measuring_drain_valve.is_closed: " +
-                            str(self.vosekast.measuring_drain_valve.is_closed))
-            self.logger.info("measuring_tank.is_filled: " +
-                            str(self.vosekast.measuring_tank.is_filled))
-
+        self.logger.info("Diagnostics:" +
+        "\nself.threads: " + str(self.threads) +
```
Comment (karlTGA) [inline vosekast_control/Scale.py]: usually the `\n` is on the end of the line and not on the beginning of the new one.

Comment (immunity-anomaly) [reply vosekast_control/Scale.py]: but then I would have to format it like this:
 "self.threads: " + str(self.threads)  + "\n" +

Comment (karlTGA) [reply vosekast_control/Scale.py]: Correct. Alternative you can use format strings:
`f"self.threads: {str(self.threads)}\n"`
Try to bring so all logs in one multiline log.

**Pull request merged**